src/HexagonOptimize.cpp

```diff
@@ -1147,6 +1219,219 @@ class OptimizePatterns : public IRMutator {
     }
 };
 
+class VectorReducePatterns : public IRMutator {
+    using IRMutator::visit;
+
+    // Match stencil patterns like reduction of a shuffle with below indices:
+    // 0, 1, 2,..., window_size - 1,
+    // 1, 2, 3,..., window_size,
+    // 2, 3, 4,..., window_size + 1
+    // .....
+    // window_size != lanes
+    // TODO: Their could be other patterns as well which we should match
+    int is_stencil_reduction(const Expr &op, int window_size) {
+        int lanes = op.type().lanes();
+        internal_assert(lanes > window_size);
+        if (const Shuffle *shuff = op.as<Shuffle>()) {
+            for (int i = window_size; i < lanes; i++) {
+                if ((i % window_size != window_size - 1) &&
+                    (shuff->indices[i - window_size + 1] != shuff->indices[i])) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    Expr visit(const Mul *op) override {
+        if (const VectorReduce *vr = op->a.as<VectorReduce>()) {
+            if (vr->op == VectorReduce::Add) {
+                // Move broadcasts inside the VectorReduce Node for
+                // vrmpy/vtmpy/vdmpy instructions.
+                const int lanes = vr->type.lanes();
+                const int arg_lanes = vr->value.type().lanes();
+                IRMatcher::Wild<0> x;
+                IRMatcher::Wild<1> y;
+                auto rewrite = IRMatcher::rewriter(IRMatcher::mul(op->a, op->b), op->type);
+                if (rewrite(h_add(x, lanes) * broadcast(y, lanes), h_add(x * broadcast(y, arg_lanes), lanes)) ||
+                    rewrite(h_add(y, lanes) * broadcast(x, lanes), h_add(broadcast(x, arg_lanes) * y, lanes))) {
```

Comment:
I don't think the second one is necessary, because a broadcast is always going to be on the right of a horizontal add (because of how the simplifier canonicalizes things)

Reply:
Also this seems risky. If you don't hit vrmpy/vtmpy/vdmpy then you just turned a narrow multiply into a much wider one.

Reply:
Will correct this in the next patch